keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,350 @@
+
+#### Provision/deletion
+Volume provision and deletion will work differently for inline volumes. For persistent inline volumes, provision/deletion is handled as follows:
+* Persistent inline volumes do not participate in the provision and deletion phases.
+* Persistent inline volumes expect provision and deletion to be handled outside of the driver.
+
+For ephemeral CSI drivers:
+* Ephemeral drivers will not receive provision/deletion API calls (since these stages are driven by a PV/PVC/StorageClass).
+* Ephemeral CSI drivers will have to delay or combine any provisioning/deprovisioning operation during different phase.
+
+#### Attach/detach
+CSI uses API object `storage.VolumeAttachment` to track and manage attach/detach storage operation. Currently that object contains reference to an associated PV when the driver is not used inline.  However,  it must be extended to support information from `CSIVolumeSource` (see above) when an inline volume is being attached.
+
+```go
+// VolumeAttachmentSpec is the specification of a VolumeAttachment request.
+type VolumeAttachmentSpec struct {
+    // <snip>
+
+	// Source represents the volume that should be attached.
+	Source VolumeAttachmentSource
+}
+
+// VolumeAttachmentSource represents a volume that should be attached, either
+// PersistentVolume or a volume in-lined in a Pod.
+// Exactly one member can be set.
+type VolumeAttachmentSource struct {
+	// Name of the persistent volume to attach.
+	// +optional
+	PersistentVolumeName *string
+
+	// InlineVolumeSource represents the source location of a in-line volume in a pod to attach.
+	// +optional
+    InlineVolumeSource *InlineVolumeSource
+}
+
+// InlineVolumeSource represents the source location of a in-line volume in a pod.
+type InlineVolumeSource struct {
+	// VolumeSource is copied from the pod. It ensures that attacher has enough
+	// information to detach a volume when the pod is deleted before detaching.
+	// Only CSIVolumeSource can be set.
+	// Required.
+	CSIVolumeSource v1.VolumeSource
+
+	// Namespace of the pod with in-line volume. It is used to resolve
+	// references to Secrets in VolumeSource.
+	// Required.
+	Namespace string
+}
+```
+The following steps outline how the API is used to track attachment and detachment of volumes:
+* The external CSI A/D controller **copies whole `VolumeSource`**  from `Pod` into `VolumeAttachment`. This allows external CSI attacher to detach volumes for deleted pods without keeping any internal database of attached VolumeSources.
+* Validation of `VolumeSource` will ensure that only `CSIVolumeSource` is being copied.  
+* External CSI attacher must be extended to  process either `PersistentVolumeName` or `VolumeSource`.
+* **External attacher may need permission to access secrets in any pod namespace** where inline volume is specified.
```

Comment:
Is this different from normal PVC/PV CSI attach secret handling?

Reply:
it is restricted in which namespaces the secrets are handled. pv's can reference any namespace. these are user controlled so should only work in their own namespace.

Reply:
does the external-attacher need to validate the namespace, or is that something kubernetes will do?

Reply:
hmm... good question. I would say kubelet? as it already has the info, and has to validate it for other places too? (drivers without attachers)

Reply:
Both Kubelet and external components must enforce this. I will update doc to reflect this security requirement.

Reply:
@msau42 that is one of the unresolved implementation detail at this point. We will need to find a workaround or change the A/D internal controller to surface that value during A/D phase.

Reply:
Or can internal A/D controller enforce it before creating the VolumeAttachment object so that the external-attacher doesn't need to care?

Reply:
@msau42 the CSI kubelet plugin A/D code can do it, but internal AD controller would need to change to bubble up namespace.

Reply:
So the namespace should be there in the volumeattachment.inlinevolumesource.  Can we validate that the namespaces in the secret ref match the pod namespace at pod creation time? Then we shouldn't need a namespace field in the volumeattachment object and the external components don't have to do extra validation

Reply:
@msau42 no, because secret refs in the `CSIVolumeSpec` uses `LocalObjectReference` (suggested by Jordan to avoid namespace injection in the pod spec).  So, the idea is to assume the secret ref will always use the namespace of their pods and the namespace is looked up during volume operation phases.

Reply:
Thanks for the clarification. That sounds fine then.  We can put the burden on the creator of the volumeattachment object to populate the namespace.  We'll need to iron out the implementation details when multiple pods in different namespaces share the same volume.

Reply:
@msau42 yeah good point - pods in different namespaces sharing the same volume is an implementation detail that came up.  Yes agreed, we'll need to find a way to do it.  Thanks for the review.